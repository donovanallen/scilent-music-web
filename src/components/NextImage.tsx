import { Skeleton } from '@nextui-org/react';
import Image, { ImageProps } from 'next/image';
import * as React from 'react';

import { cn } from '@/lib/utils';

type NextImageProps = {
  useSkeleton?: boolean;
  classNames?: {
    image?: string;
    blur?: string;
  };
  alt: string;
} & (
  | { width: string | number; height: string | number }
  | { layout: 'fill'; width?: string | number; height?: string | number }
  | { fill: boolean }
) &
  ImageProps;

/**
 *
 * @description Must set width using `w-` className
 * @param useSkeleton add background with pulse animation, don't use it if image is transparent
 */
export default function NextImage({
  useSkeleton = false,
  src,
  width,
  height,
  fill,
  alt,
  className,
  classNames,
  ...rest
}: NextImageProps) {
  const [isImageLoaded, setIsImageLoaded] = React.useState(false);
  const [_status, _setStatus] = React.useState(
    useSkeleton ? 'loading' : 'complete',
  );
  const widthIsSet = className?.includes('w-') ?? false;

  return (
    <Skeleton className='rounded-md bg-neutral-500' isLoaded={isImageLoaded}>
      <div
        style={!widthIsSet ? { width: `${width}px` } : undefined}
        className={className}
      >
        <Image
          className={cn(
            'relative aspect-square object-cover',
            classNames?.image,
          )}
          src={src}
          width={width}
          height={height}
          fill={fill}
          alt={alt}
          onLoad={() => setIsImageLoaded(true)}
          sizes='(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw'
          {...rest}
        />
      </div>
    </Skeleton>
  );
}
