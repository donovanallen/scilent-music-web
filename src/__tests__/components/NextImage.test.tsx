import { render } from '@testing-library/react';
import React from 'react';

import NextImage from '@/components/NextImage';

describe('NextImage', () => {
  const mockImageSrc = 'http://www.example.com';
  it('renders correctly', () => {
    const { asFragment } = render(
      <NextImage alt='test' fill src={mockImageSrc} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
