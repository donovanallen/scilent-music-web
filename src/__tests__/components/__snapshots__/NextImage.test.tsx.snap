// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextImage renders correctly 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="test"
      class="aspect-square object-cover"
      data-nimg="fill"
      decoding="async"
      loading="lazy"
      sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
      src="/_next/image?url=http%3A%2F%2Fwww.example.com&w=3840&q=75"
      srcset="/_next/image?url=http%3A%2F%2Fwww.example.com&w=256&q=75 256w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=384&q=75 384w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=640&q=75 640w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=750&q=75 750w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=828&q=75 828w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=1080&q=75 1080w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=1200&q=75 1200w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=1920&q=75 1920w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=2048&q=75 2048w, /_next/image?url=http%3A%2F%2Fwww.example.com&w=3840&q=75 3840w"
      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
    />
  </div>
</DocumentFragment>
`;
